Develop/ram console option known addr buf

Zhiqiang-Hou opened the pull request:
This PR added an option for RAM console driver to locate the console buffer to a predefined memory region, so that the RAM console can be used easily, and also added an example on imx8mmevk.

Zhiqiang-Hou committed:
arm64: linker: create region and section from zephyr,memory-region

In arm64 linker script, create a memory region and section for each
device tree node with compatible string "zephyr,memory-region".

Signed-off-by: Hou Zhiqiang <Zhiqiang.Hou@nxp.com>

Zhiqiang-Hou committed:
drivers: ram_console: add option to link console buffer to given section

Currently, the ram_console buffer is defined as a global var, its
address is determined during the building and may be changed when
code update. This is not a problem if the ram_console is just used
in debug purpose.

While in the heterogeneous SoCs, there can be multiple MPU Cores
and several MCU cores, it can run multiple OS/bare-metal instances
on these cores, but the UART ports may be not enough, so the
ram_console can be leveraged. To make it easy to use, it's better
make the console buffer fixed and predefined.

This patch adds a option to link the console buffer to a given
section, through the "zephyr,memory-region" device tree node, then
the address can be known from the device tree node and easy to
check from other cores running Linux/U-Boot.

To use this option, the chosen property 'zephyr,ram-console' must
be added, the following is a example:

	chosen {
		zephyr,ram-console = &ram_console;
	};

	ram_console: memory@93d00000 {
		compatible = "zephyr,memory-region";
		reg = <0x93d00000 DT_SIZE_K(4)>;
		zephyr,memory-region = "RAM_CONSOLE";
	};

Signed-off-by: Hou Zhiqiang <Zhiqiang.Hou@nxp.com>

Zhiqiang-Hou committed:
soc: imx9: a55: create region and section from the zephyr,memory-region

It has been added in the arm64 common linker script, so also update
the one of mimx9 SoCs.

Signed-off-by: Hou Zhiqiang <Zhiqiang.Hou@nxp.com>

Zhiqiang-Hou commented:
This update fixed some check failures.

carlocaione commented on drivers/console/ram_console.c:
```diff
@@ -11,23 +12,36 @@
 extern void __printk_hook_install(int (*fn)(int));
 extern void __stdout_hook_install(int (*fn)(int));
 
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(ram_console), okay)
```
what about using a `chosen` node instead?

Zhiqiang-Hou replied:
Will update. Thanks a lot for your comments!

Zhiqiang-Hou replied:
updated, please help to review again. Thanks!

andyross commented on drivers/console/ram_console.c:
```diff
@@ -11,23 +12,36 @@
+#if DT_HAS_CHOSEN(zephyr_ram_console)
+	size = DT_REG_SIZE(DT_CHOSEN(zephyr_ram_console)) - 1;
+#else
+	size = CONFIG_RAM_CONSOLE_BUFFER_SIZE;
```
Not a big deal, but note that both of these values which can be assigned to size are compile-time constants, so you're adding needless bytes to .data and an extra load per character output when you could have just done this using the preprocessor.

Zhiqiang-Hou replied:
Thanks for your comments!
Updated to use preprocessor.

Zhiqiang-Hou commented:
Updated to fix LONG_LINE check warning.

dleach02 reviewed (changes requested):
This will conflict with the HWMv2 so we will block this till the collab branch merges

Zhiqiang-Hou commented:
> This will conflict with the HWMv2 so we will block this till the collab branch merges

Okay, will rebase then.

nordicjm reviewed (changes requested):
Needs converting to hwmv2

Zhiqiang-Hou commented:
> Needs converting to hwmv2

Yes, will rebase after the platform hwmv2 migration PR merged.

dleach02 commented:
@Zhiqiang-Hou HWMv2 has merged into main. Please rebase and adjust your PR.

Zhiqiang-Hou commented:
> @Zhiqiang-Hou HWMv2 has merged into main. Please rebase and adjust your PR.

Thanks for the reminding! Will rebase later.

Zhiqiang-Hou commented:
rebased to hwmv2 and added imx8mpevk, imx8mnevk and imx93evk A-core RAM console variant.
based on https://github.com/zephyrproject-rtos/zephyr/pull/67886

nordicjm commented on boards/nxp/imx8mm_evk/imx8mm_evk_mimx8mm6_a53_ramconsole.dts:
```diff
@@ -0,0 +1,42 @@
+	compatible = "fsl,mimx8mm";
+
+	chosen {
+		zephyr,sram = &sram0;
+		zephyr,ram-console = &ram_console;
```
Any particular reason this wouldn't be a snippet instead?

Zhiqiang-Hou replied:
This is suggested by Carlo, refer to https://github.com/zephyrproject-rtos/zephyr/pull/68452#discussion_r1477925702

nordicjm replied:
If this is a snippet then the existing board targets (including variants) can be used as they are, and optionally switched to RAM console by doing e.g. `-DSNIPPET=imx8mm_ram_console` or similar. Thoughts @gmarull @tejlmand ?

Zhiqiang-Hou replied:
I misunderstood your previous comment, I didn't know the snippet in Zephyr.
It's a good idea to use the snippet. Thanks for the suggestion!

nordicjm commented on boards/nxp/imx8mm_evk/imx8mm_evk_mimx8mm6_a53_ramconsole_defconfig:
```diff
@@ -0,0 +1,25 @@
+# SPDX-License-Identifier: Apache-2.0
+
+# ARM Options
+CONFIG_AARCH64_IMAGE_HEADER=y
```
Any option here that exists in `imx8mm_evk_mimx8mm6_a53_defconfig` is not needed, applies to the other commits too

Zhiqiang-Hou replied:
Indeed, thanks!

Zhiqiang-Hou replied:
updated.

JiafeiPan commented:
It seems some commits in another PR https://github.com/zephyrproject-rtos/zephyr/pull/67886 is pushed to this PR.

JiafeiPan commented on boards/nxp/imx93_evk/imx93_evk_mimx9352_a55_ramconsole.dts:
```diff
@@ -0,0 +1,108 @@
+/dts-v1/;
+
+#include <nxp/nxp_mimx93_a55.dtsi>
+#include "imx93_evk-pinctrl.dtsi"
```
Most of dts nodes are the same with imx93_evk_mimx9352_a55.dts, how about include imx93_evk_mimx9352_a55.dts directly? then only add updates in this file, it will be easy to maintain.

Zhiqiang-Hou replied:
As Jamie suggested the snippet, will try to convert to snippet, so just add overlay dts. Thanks!

Zhiqiang-Hou commented on soc/nxp/imx/imx8m/a53/mmu_regions.c:
```diff
@@ -35,6 +35,18 @@ static const struct arm_mmu_region mmu_regions[] = {
+			      MT_DEVICE_nGnRnE | MT_P_RW_U_NA | MT_NS),
+
+#if DT_NODE_HAS_STATUS(DT_CHOSEN(zephyr_ram_console), okay)
+	MMU_REGION_FLAT_ENTRY("RAM_CONSOLE",
```
Make sense, will rework the patches. Thanks!

Zhiqiang-Hou committed:
snippets: add ram-console snippet support

The ram-console snippet disabled UART console and enabled the
RAM console with the option link the RAM console buffer to a
dedicate section in a new added memory-region.

Signed-off-by: Hou Zhiqiang <Zhiqiang.Hou@nxp.com>

Zhiqiang-Hou committed:
driver: ram_console: leave one byte from the defined buffer size

Leave one byte from the CONFIG_RAM_CONSOLE_BUFFER_SIZE to ensure
the NULL-termination.

Signed-off-by: Hou Zhiqiang <Zhiqiang.Hou@nxp.com>

JiafeiPan commented on drivers/console/ram_console.c:
```diff
@@ -11,23 +12,47 @@
+	mm_reg_t ram_console_va;
+
+	device_map((mm_reg_t *)&ram_console_va, DT_REG_ADDR(DT_CHOSEN(zephyr_ram_console)),
+		   CONFIG_RAM_CONSOLE_BUFFER_SIZE, K_MEM_CACHE_NONE | K_MEM_DIRECT_MAP);
```
How about get the RAM Console buffer size from DT_REG_SIZE()? otherwise there will be issue if memory region size configured in dts lower than CONFIG_RAM_CONSOLE_BUFFER_SIZE.

Zhiqiang-Hou replied:
Thanks for your comments!
Already added a check to ensure the memory region size won't less than CONFIG_RAM_CONSOLE_BUFFER_SIZE:
```
#elif (CONFIG_RAM_CONSOLE_BUFFER_SIZE > DT_REG_SIZE(DT_CHOSEN(zephyr_ram_console)))
#error "Customed RAM console buffer exceeds the section size!"
#endif
```

Zhiqiang-Hou commented:
updates:
1. Changed to add a `ram-console` Snippet, so that other platforms/variants can easily enable this feature.
2. Added a Config entry to control if the RAM console buffer uses a dedicated memory-region/section.
3. Moved the RAM console buffer memory mapping to ram-console driver.
4. Changed the CONFIG_RAM_CONSOLE_BUFFER_SIZE as the total buffer size and leaved one byte internal to ensure the NULL-termination.

JiafeiPan commented on snippets/ram-console/ram-console.conf:
```diff
@@ -0,0 +1,9 @@
+# Enable Console
+# CONFIG_UART_CONSOLE is not set
+CONFIG_RAM_CONSOLE=y
+CONFIG_RAM_CONSOLE_BUFFER_SECTION=y
```
CONFIG_KERNEL_DIRECT_MAP=y as it is used in ram_console.c

Zhiqiang-Hou replied:
it is selected by the CONFIG_RAM_CONSOLE_BUFFER_SECTION. Thanks!

dleach02 commented:
@nordicjm can you revisit this PR

nordicjm commented on snippets/ram-console/ram-console.conf:
```diff
@@ -0,0 +1,9 @@
+# SPDX-License-Identifier: Apache-2.0
+
+# Enable Console
+# CONFIG_UART_CONSOLE is not set
```
```suggestion
```
Don't need a line saying not set, if you want to turn it off you need to have it `=n` here

Zhiqiang-Hou replied:
I want to disable UART console, will update it to `=n`. Thanks!

Zhiqiang-Hou commented:
> @nordicjm can you revisit this PR

Sure and thanks for your review and approval!

kartben commented on snippets/ram-console/README.rst:
```diff
@@ -0,0 +1,29 @@
+How to add support of a new board
+*********************************
+
+Add board dts overlay to this snippet to add memory-region node with node
+label ``snippet_ram_console`` as the following:
```
I am not sure this is how we want folks to use snippets, i.e. a snippet should work "out of the box" without being board specific. In this particular case of course, the presence of the DT node is required but I'm not sure adding an overlay for it *in the snippet* is what we want to recommend?

Zhiqiang-Hou replied:
Updated the PR to make it work "out of the box":
Without board specific overlay, it enables the RAM console using a global array for RAM console buffer.

Zhiqiang-Hou commented on drivers/console/Kconfig:
```diff
@@ -104,13 +104,28 @@ config RAM_CONSOLE
+	help
+	  Use a dedicated section as the RAM console buffer, whose address is
+	  known before build so that the console output messages can be easily
+	  examined by a debuffer or software tool from a parallel-running OS.
```
Fixed the typo in the last update. Thanks!

Zhiqiang-Hou commented on drivers/console/ram_console.c:
```diff
@@ -11,23 +12,47 @@
+#if !DT_HAS_CHOSEN(zephyr_ram_console)
+#error "Lack of chosen property zephyr,ram_console!"
+#elif (CONFIG_RAM_CONSOLE_BUFFER_SIZE > DT_REG_SIZE(DT_CHOSEN(zephyr_ram_console)))
+#error "Customed RAM console buffer exceeds the section size!"
```
Yes, will fix.

Zhiqiang-Hou commented on snippets/ram-console/README.rst:
```diff
@@ -0,0 +1,29 @@
+
+This snippet redirects console output to a RAM buffer, which is linked to
+a dedicated section created by linker scripts through a device tree node
+with compatible string "zephyr,memory-region".
```
Thanks for the suggestion! will update.

nordicjm commented on drivers/console/Kconfig:
```diff
@@ -109,8 +109,10 @@ config RAM_CONSOLE_BUFFER_SIZE
-	  length is exceeded.
+	  Total size of the RAM console buffer, to ensure it's always
+	  NULL-terminated leave one byte unused, the actual length is
+	  less one byte. Messages will wrap around if the actual length
```
*one byte less

Zhiqiang-Hou replied:
will update. Thanks for correcting!

dleach02 commented:
@kartben can you review the changes to your change request
@carlocaione you had a previous approval but that was dropped to address a different request.

kartben commented on snippets/ram-console/README.rst:
```diff
@@ -0,0 +1,37 @@
+********
+
+This snippet redirects console output to a RAM buffer. The RAM console
+buffer is a global array lacated in RAM region by default, whose address
```
```suggestion
buffer is a global array located in RAM region by default, whose address
```

kartben commented on snippets/ram-console/README.rst:
```diff
@@ -0,0 +1,37 @@
+.. code-block:: DTS
+
+    / {
+        chosen {
+            zephyr,ram-console = &snippet_ram_console;
+        };
+
```
Instead of repeating this in each board overlay, it would probably make sense to move it to a common `ram-console.overlay`

Zhiqiang-Hou replied:
By default, the RAM console uses a global array for the console buffer, it doesn't need this chosen property. It is only needed when the RAM console wants to use a dedicated section for console buffer.

kartben commented:
@carlocaione could you please refresh your +1? thanks!
